Support `_FILE` suffixed env vars in Docker entrypoint

pugnascotia opened the pull request:
Closes #43603. Allow environment variables to be passed to ES in a Docker container via a file, by setting an environment variable with the `_FILE` suffix that points to the file with the intended value of the env var.

For example, the following command runs an ES container and sets the bootstrap password via the `/run/secrets/bootstrapPassword.txt` file, by bind-mounting a directory as `/run/secrets/`.

```
docker run \
  -it \
  --rm \
  -p 9200:9200 \
  -p 9300:9300 \
  -e ELASTIC_PASSWORD_FILE=/run/secrets/bootstrapPassword.txt \
  -v $PWD/secrets:/run/secrets \
  -e "discovery.type=single-node" \
  -e xpack.security.enabled=true \
  elasticsearch:test
```

pugnascotia committed:
Support ELASTIC_PASSWORD_FILE in Docker entrypoint

Closes #43603. Allow a password to be specifed in an ES Docker container
by specifying a file path in the ELASTIC_PASSWORD_FILE environment
variable.

pugnascotia committed:
Support all envs vars with a _FILE suffix

Expand support for populating environment variables from file via vars
with a _FILE suffix.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Packaging)

alpar-t reviewed (commented):
We should add some tests to verify this behavior.

atorok commented on distribution/docker/src/docker/bin/docker-entrypoint.sh:
```diff
@@ -38,6 +38,38 @@ if [[ "$1" != "eswrapper" ]]; then
+    fi
+
+    if [[ ! -e "${!VAR_NAME_FILE}" ]]; then
+      echo "ERROR: File ${!VAR_NAME_FILE} does not exist" >&2
```
May be worth extending the message to also include the variable name pointing to that file

atorok commented on distribution/docker/src/docker/bin/docker-entrypoint.sh:
```diff
@@ -38,6 +38,38 @@ if [[ "$1" != "eswrapper" ]]; then
+    fi
+
+    echo "Setting $VAR_NAME from ${!VAR_NAME_FILE}" >&2
+    export "$VAR_NAME"="$(cat ${!VAR_NAME_FILE})"
```
Should we be stripping a possible newline at the end of the file here ?

pugnascotia replied:
The newline is already stripped thanks to `bash`. An earlier implementation read the file without using `cat`, but didn't strip the newline.

pugnascotia committed:
Include _FILE names when checking files in Docker

In the Docker entrypoint, when using _FILE-suffixed env vars to populate
environment variables, include the _FILE var name for context.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 43603-docker-entrypoint-file-suffix

pugnascotia committed:
Trying to write tests

pugnascotia committed:
Fix up new Docker test

pugnascotia committed:
Revert accidental changes

pugnascotia committed:
Tweaks

pugnascotia commented:
@atorok thanks for the review, I've addressed your points. Of course, the new test doesn't run yet in CI due to #47639, so I've been running it locally.

alpar-t reviewed (commented):
Thanks this LGTM now. 

I'm wondering if we should enforce safe permissions on these files. 
If i'm reading this right, the env vars and thus the contents of the file, 
could be used to perform command injection in the container. 
Files make this worse,  because all it takes is for the file to be writable by someone else 
and it could trick the administrator into running those commands, granted in the container but that might still be bad enough as one could access the data directory.

atorok commented:
Suppose a file `value.txt` exists and we have some FILE_SETTING variable configured to point to it.  
If `value.txt` is word writable, one could edit it to something like 
```
some value" && rm -Rf /path/to/data/dir # 
```
So this will eventually run something like :
```
run_as_other_user_if_needed /usr/share/elasticsearch/bin/elasticsearch "-Esetting=some value" && rm -Rf /path/to/data/dir # any other arguments here don't matter  
```

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging

pugnascotia commented:
I'm waiting to merge #47640 before finishing this off.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 43603-docker-entrypoint-file-suffix

pugnascotia committed:
Post-merge fixes

pugnascotia committed:
Check env var file permissions

Require files that are being used to populate env vars to have
restricted file permissions. This is to attempt to limit the possibility
of crafting a malicious command line for ES.

pugnascotia commented:
GitHub seems to have forgotten that this PR needs CI. I'll try closing and reopening.

pugnascotia commented:
Oh, GitHub's having issues, I'll keep an eye on it.

@atorok I've tried to restrict the permissions on files being used for env var. They have to be mode 600 (like an SSH private key) or 400 (even tighter), but I'd love feedback / opinions on this. Please take a look.

@dliappis I'd like your feedback on this PR too.

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix

pugnascotia committed:
Capture more logging on failure

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix to pick up the right commit

pugnascotia committed:
Split env var file tests

The test test80SetEnvironmentVariablesUsingFiles relied on x-pack
security being available, which isn't the case in the OSS image.

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix

pugnascotia commented:
@elasticmachine upgrade branch

pugnascotia commented:
@elasticmachine update branch, I meant to say

pugnascotia commented:
@elasticmachine update branch please?

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 43603-docker-entrypoint-file-suffix

atorok commented on distribution/docker/src/docker/bin/docker-entrypoint.sh:
```diff
@@ -38,6 +38,45 @@ if [[ "$1" != "eswrapper" ]]; then
+      exit 1
+    fi
+
+    if [[ ! -r "${!VAR_NAME_FILE}" ]]; then
```
Might be worth combining the checks on permissions into a single  one.
We could get into the situation when we tell the user it's not readable, then we tell him it's too broadly readable.

atorok commented on distribution/docker/src/docker/bin/docker-entrypoint.sh:
```diff
@@ -38,6 +38,45 @@ if [[ "$1" != "eswrapper" ]]; then
+    FILE_PERMS="$(stat -c '%a' ${!VAR_NAME_FILE})"
+
+    if [[ "$FILE_PERMS" != "400" && "$FILE_PERMS" != 600 ]]; then
+        echo "ERROR: File ${!VAR_NAME_FILE} from $VAR_NAME_FILE must have file permissions 400 or 600" >&2
```
Maybe including what permissions it actually has might be useful  to the user sometimes ? 

The permissions are a bit on the strict size. We could technically be asserting that no write permissions are available to anyone besides the user, but since it's expected that these files will mostly be used for secrets I guess this is fine.

atorok commented on qa/os/src/test/java/org/elasticsearch/packaging/test/DockerTests.java:
```diff
@@ -178,6 +176,122 @@ public void test70BindMountCustomPathConfAndJvmOptions() throws Exception {
+     * which point to files that hold the required values.
+     */
+    public void test80SetEnvironmentVariablesUsingFiles() throws Exception {
+        final Path secretsDir = getTempDir().resolve("secrets");
```
I would pull this into an instance variable and use `@Before` and `@After` to set it and clean up the dir. 
It's more readable and less repetitive avoiding the try, finally is particularly beneficial imho.

atorok commented on qa/os/src/test/java/org/elasticsearch/packaging/test/DockerTests.java:
```diff
@@ -178,6 +176,122 @@ public void test70BindMountCustomPathConfAndJvmOptions() throws Exception {
+            runContainer(distribution(), volumes, envVars);
+
+            // If we configured security correctly, then this call will only work if we specify the correct credentials.
+            waitForElasticsearch("green", null, installation, "elastic", "hunter2");
```
We might want to wrap the exception here to add some additional context.
Imagine this fails 6 months from now, will it be obvious from the failure that the passwrod was not applied or will it look as if elasticsearch didn't start ? I'm not sure what the error looks like, guess it's better if it's an authentication failure, but the additional context certainly doesn't hurt.

alpar-t reviewed (approved):
LGTM overall. I left some comments that I think should be looked at, but no need for an additional review.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 43603-docker-entrypoint-file-suffix

pugnascotia committed:
Address review feedback

dliappis commented on distribution/docker/src/docker/bin/docker-entrypoint.sh:
```diff
@@ -38,6 +38,40 @@ if [[ "$1" != "eswrapper" ]]; then
+    VAR_NAME="${VAR_NAME_FILE%_FILE}"
+
+    if [[ -n "${!VAR_NAME}" ]]; then
+      echo "ERROR: Both $VAR_NAME_FILE and $VAR_NAME are set. These are mutually exclusive." >&2
```
As this is now, this conditional will be skipped if an ENV VAR (say ...`ELASTIC_PASSWORD`) is set to empty content (by accident, for example).

Our code later on will still present this empty env var as `-E ...` cli arg.

What if instead of checking the length of the content of `VAR_NAME` we just ensure it's defined as an env var?

This can be achieved with `env | cut -f1 -d= | grep ^$VAR_NAME\$ >/dev/null` similar to what you've done earlier, or just with `printenv $VAR_NAME >/dev/null`. Both will fail if `VAR_NAME` doesn't match .

dliappis replied:
https://github.com/elastic/elasticsearch/pull/47573/commits/0a07f70223ecd4c8e8061638277f2abdb52c75c0 that resolves this comment looks good, thanks.

dliappis reviewed (commented):
Very nicely done! I left a minor comment up for discussion.

pugnascotia committed:
Tighter checks for _FILE vs regular vars

Ensure that regular environment variables are not even set at the same
time as their _FILE equivalents. Previously they could be set so long as
they were empty.

pugnascotia committed:
Update docs for _FILE variables

jrodewig commented on docs/reference/setup/install/docker.asciidoc:
```diff
@@ -312,7 +312,16 @@ over the configuration files in the image.
 You can set individual {es} configuration parameters using Docker environment variables. 
 The <<docker-compose-file, sample compose file>> and the 
-<<docker-cli-run-dev-mode, single-node example>>  use this method. 
+<<docker-cli-run-dev-mode, single-node example>>  use this method.
```
```suggestion
<<docker-cli-run-dev-mode, single-node example>> use this method.
```

jrodewig commented on docs/reference/setup/install/docker.asciidoc:
```diff
@@ -312,7 +312,16 @@ over the configuration files in the image.
-<<docker-cli-run-dev-mode, single-node example>>  use this method. 
+<<docker-cli-run-dev-mode, single-node example>>  use this method.
+
+If you are providing secrets e.g. passwords to {es} and you do not want to pass them
```
```suggestion
If you are providing secrets, such as passwords, to {es} and do not want to pass them
```

jrodewig commented on docs/reference/setup/install/docker.asciidoc:
```diff
@@ -312,7 +312,16 @@ over the configuration files in the image.
+<<docker-cli-run-dev-mode, single-node example>>  use this method.
+
+If you are providing secrets e.g. passwords to {es} and you do not want to pass them
+directly via environment variables, you can instead supply environment variables suffixed with
```
```suggestion
directly via environment variables, you can supply environment variables suffixed with
```

jrodewig commented on docs/reference/setup/install/docker.asciidoc:
```diff
@@ -312,7 +312,16 @@ over the configuration files in the image.
+
+If you are providing secrets e.g. passwords to {es} and you do not want to pass them
+directly via environment variables, you can instead supply environment variables suffixed with
+"_FILE". The variable value specifies a file, whose contents are used for the secret.
```
```suggestion
"_FILE". The variable value specifies a file whose contents are used for the secret.
```

jrodewig commented on docs/reference/setup/install/docker.asciidoc:
```diff
@@ -312,7 +312,16 @@ over the configuration files in the image.
+"_FILE". The variable value specifies a file, whose contents are used for the secret.
+
+For example, if you specified the environment variable "ELASTIC_PASSWORD_FILE" with value
+"/run/secrets/password.txt", and bind-mounted a file at this location in the container with
```
```suggestion
"/run/secrets/password.txt" and bind-mounted a file at this location in the container with
```

jrodewig commented on docs/reference/setup/install/docker.asciidoc:
```diff
@@ -312,7 +312,16 @@ over the configuration files in the image.
+
+For example, if you specified the environment variable "ELASTIC_PASSWORD_FILE" with value
+"/run/secrets/password.txt", and bind-mounted a file at this location in the container with
+the contents "PleaseChangeMe", then this would be equivalent to specifying the environment
```
```suggestion
the contents "PleaseChangeMe", this would be equivalent to specifying the environment
```

pugnascotia committed:
Docs tweaks

debadair commented on docs/reference/setup/install/docker.asciidoc:
```diff
@@ -312,7 +312,16 @@ over the configuration files in the image.
+For example, if you specified the environment variable "ELASTIC_PASSWORD_FILE" with value
+"/run/secrets/password.txt" and bind-mounted a file at this location in the container with
+the contents "PleaseChangeMe", this would be equivalent to specifying the environment
+variable "ELASTIC_PASSWORD" with the value "PleaseChangeMe".
```
In general, I'd recommend sticking to present tense for examples. In this case, I'd go with something like:

> For example, to set the {es} bootstrap password from a file, you can bind mount the file and set the `ELASTIC_PASSWORD_FILE` environment variable to the mount location. If you mount the password file to `/run/secrets/password.txt`, specify:
> 
> `-e ELASTIC_PASSWORD_FILE=/run/secrets/bootstrapPassword.txt \`

Also note that the quotes should be backticks to render the env var and path in code font.

debadair reviewed (approved):
A couple editorial comments, but LGTM.

pugnascotia committed:
Address further docs review feedback

pugnascotia commented:
Thanks @debadair, I've incorporated your feedback.

dliappis reviewed (approved):
LGTM thanks

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 43603-docker-entrypoint-file-suffix

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix
